### DISCUSSION TITLE
Repeatable content for simple property editors

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#17566] [Features and ideas]
Sometimes we need to control image (or other content) specific to responsive layout, e.g. Mobile, Tablet and Desktop.
We could add three properties "Image Mobile", "Image Tablet" and "Image Desktop", but sometimes I think it would be simpler with a single property "Image" and then defining breakpoint configuration in property editor, wrapping an existing property editor.

With image it can be solved using different crops as long it is same image, but we have sometimes has the need to control an embed, video, text or similar per breakpoint (and eventually used device detection in code).

Nested Content / Block List could be used, e.g. adding three blocks, but it would probably need to be three different blocks to know its type and if it is for Mobile, Tablet or Desktop.

Perhaps a combination of repeatable textstrings (perhaps repeatable numbers, dates etc. in future) and Nested Content / Block List.

An example could be something like this.
Perhaps listed as tabs with icon/name + textbox under each tab panel.

```
[
   {
       "alias": "mobile",
       "icon": "icon-mobile",
       "name": "Mobile",
       "value": "This is a text on mobile"
   },
   {
       "alias": "tablet",
       "icon": "icon-tablet",
       "name": "Tablet",
       "value": "This is a text on tablet"
   },
   {
       "alias": "desktop",
       "icon": "icon-desktop",
       "name": "Desktop",
       "value": "This is a text on desktop"
   }
}
```

Not necessary just for breakpoints, but could be anything where one want to define and key/alias and connect it with a specific value. 

[Vorto](https://our.umbraco.com/packages/backoffice-extensions/vorto/) did it it some way connecting a value to a culture and similar for [Multilanguage Textbox](https://marketplace.umbraco.com/package/our.umbraco.multilanguagetextbox)

In these use-cases I think it would be a bit overkill with Block List and not easy to target a specific entry/item unless one add a key/alias to each item, but often we don't want content editors to worry about this.

Key/alias could be: 
Culture/Language code, user-defined alias, etc. The culture/language code could be hard-coded, but may be useful to select existing BE languages.

Value could be:
String, number (integer/decimal/float), datetime.

**Comment by skttl** [comment]
Could be a combination of these package ideas
https://github.com/leekelleher/umbraco-package-ideas/issues/47
https://github.com/leekelleher/umbraco-package-ideas/issues/48

**Comment by leekelleher** [comment]
From a community perspective, I'd started going down this route with the [Textbox List editor in Contentment](https://github.com/leekelleher/umbraco-contentment/blob/dev/v6.x/docs/editors/textbox-list.md), but then ended up sticking with textbox inputs _(probably for simplicity and my own specific needs at the time)._

I definitely like the idea of a repeatable tuple-type editor, could be quite exciting! 😀 
Probably best prototyped as a community extension first, for exploration/iteration, etc.

**Reply by bjarnef** [reply]
@leekelleher perhaps it is easier to re-use input component and input type in the new backoffice.
Often a repeatable textstring array, but could also be list of `email`, `url`, `number` (int/decimal/float), `date` input fields.
I guess it could possible make the existing repeatable textstring property editor obsolete at some point.

**Comment by bjarnef** [comment]
Instead if `name` property, `label` may be better and if we sometime get label/value option for dropdown, checkbox/radiobutton property editors.

**Comment by bjarnef** [comment]
@iOvergaard @leekelleher I wonder if we a simple approach could be to add a configuration option to the existing repeatable textstrings property editor allowing to select input type supported by `uui-input`.
With this is can possible be based on `InputMode` type: https://github.com/umbraco/Umbraco.UI/pull/1021

I don't see a use- case for `search` and `password` input type though. `email`, `tel` and `url` shouldn't effect the appearance, but perhaps it would be nice with client side and possible server side validation.

And numeric values would probably need validation as well.

It may only need pattern, e.g. `number` input is shown a bit different on iOS and Android, but combined with `pattern="[0-9]*"` virtual keypad would show slighty different (on iOS or Android), but with `inputmode` I think it can be used instead.

Although we can use blocks, sometimes a repeatable property is simpler. E.g. to add more than one email or phone number for an employee (or URLs for social media - in that case an optional icon + alias would be useful to easy identify item from array https://github.com/umbraco/Umbraco-CMS/discussions/17566#discussion-7505781 )

**Comment by bjarnef** [comment]
I think this is still relevant and perhaps something, which possible could be used for recipient email address(es) in Umbraco Forms or similar use case, but since multiple addresses can be added separated by semicolons or commas, it has for now used textbox instead of email address property editor:
https://docs.umbraco.com/umbraco-forms/editor/attaching-workflows/workflow-types#send-email

<img width="784" height="1196" alt="image" src="https://github.com/user-attachments/assets/6d5a45a4-f885-4b3b-9ba7-256d96ea17e4" />

In use cases where only an array of `string`, `int`, `decimal` etc. I think a block list would complicate things more than necessary.
